backend/gce.go

```diff
@@ -364,18 +381,30 @@ func newGCEProvider(c ConfigGetter) (Provider, error) {
 		uploadRetries:     uint64(c.Int("upload-retries")),
 		uploadRetrySleep:  c.Duration("upload-retry-sleep"),
 
-		rateLimiter: time.NewTicker(c.Duration("rate-limit-tick")),
+		rateLimiter:       ratelimit.NewRateLimiter(redisConn, c.String("rate-limit-prefix")),
+		rateLimitMaxCalls: uint64(c.Int("rate-limit-max-calls")),
+		rateLimitDuration: c.Duration("rate-limit-duration"),
 	}, nil
 }
 
 func (p *gceProvider) apiRateLimit() {
-	atomic.AddUint64(&p.rateLimitQueueDepth, 1)
 	metrics.Gauge("travis.worker.vm.provider.gce.rate-limit.queue", int64(p.rateLimitQueueDepth))
 	startWait := time.Now()
-	<-p.rateLimiter.C
-	metrics.TimeSince("travis.worker.vm.provider.gce.rate-limit", startWait)
+	defer metrics.TimeSince("travis.worker.vm.provider.gce.rate-limit", startWait)
+
+	atomic.AddUint64(&p.rateLimitQueueDepth, 1)
 	// This decrements the counter, see the docs for atomic.AddUint64
-	atomic.AddUint64(&p.rateLimitQueueDepth, ^uint64(0))
+	defer atomic.AddUint64(&p.rateLimitQueueDepth, ^uint64(0))
+
+	for {
+		ok, err := p.rateLimiter.RateLimit("gce-api", p.rateLimitMaxCalls, p.RateLimitDuration)
```

Comment:
An error from the rate limiter is treated no differently than "rate limiter said 'no'" right now, is that what we want or should they be treated differently?